ckanapi/datapackage.py

```diff
@@ -50,69 +50,93 @@ def create_datapackage(record, base_path, stderr):
     datapackage_dir = os.path.join(base_path, dataset_name)
     os.makedirs(os.path.join(datapackage_dir, 'data'))
 
-    resources = []
+    # filter out some resources
     ckan_resources = []
     for resource in record.get('resources', []):
         if resource['format'] in resource_formats_to_ignore:
             continue
-        resources.append(create_resource(resource, datapackage_dir, stderr))
         ckan_resources.append(resource)
+    dataset = dict(record, resources=ckan_resources)
 
-    json_path = os.path.join(datapackage_dir, 'datapackage.json')
-    datapackage = dataset_to_datapackage(dict(record, resources=resources))
+    # get the datapackage (metadata)
+    datapackage = dataset_to_datapackage(dataset)
 
-    # prefer resource names from datapackage metadata
+    existing_filenames = []
     for cres, dres in zip(ckan_resources, datapackage.get('resources', [])):
-        name = dres['name']
-        ext = slugify.slugify(dres['format'])
-        if name.endswith(ext):
-            name = name[:-len(ext)]
-        try:
-            os.rename(
-                os.path.join(datapackage_dir, 'data', cres['id']),
-                os.path.join(datapackage_dir, 'data', name + '.' + ext))
-            # successful local download
-            dres['path'] = 'data/' + name + '.' + ext
-        except OSError:
-            pass
+        filename = resource_filename(dres, existing_filenames)
 
-        # convert datastore data dictionary to datapackage schema
-        if 'schema' not in dres and 'datastore_fields' in cres:
-            fields = []
-            for f in cres['datastore_fields']:
-                if f['id'] == '_id':
-                    continue
-                df = {'name': f['id']}
-                dtyp = DATAPACKAGE_TYPES.get(f['type'])
-                if dtyp:
-                    df['type'] = dtyp
-                dtit = f.get('info', {}).get('label', '')
-                if dtit:
-                    df['title'] = dtit
-                ddesc = f.get('info', {}).get('notes', '')
-                if ddesc:
-                    df['description'] = ddesc
-                fields.append(df)
-            dres['schema'] = {'fields': fields}
+        # download the resource
+        cres = \
+            create_resource(resource, filename, datapackage_dir, stderr)
+        dres['path'] = 'data/' + filename
 
+        populate_schema_from_datastore(cres, dres)
 
+    json_path = os.path.join(datapackage_dir, 'datapackage.json')
     with open(json_path, 'wb') as out:
         out.write(pretty_json(datapackage))
 
+    return datapackage_dir, datapackage, json_path
+
 
-def populate_datastore_fields(ckan, dataset):
+def resource_filename(dres, existing_filenames):
+    # prefer resource names from datapackage metadata
+    name = dres['name']
+    ext = slugify.slugify(dres['format'])
+    if name.endswith(ext):
+        name = name[:-len(ext)]
+    # deduplicate
```

Comment:
the datapackage code already deduplicates the names, no?

Reply:
https://github.com/ckan/ckanapi/blob/d3a7990230c15d69665990501e89a703bafa283f/ckanapi/datapackage.py#L196-L204

Reply:
agreed, I'll remove this